File: src/ragrep/ingest/pipeline.py
"""Ingestion pipeline orchestrator."""

import hashlib
import logging
import pickle
import time
from pathlib import Path

import faiss
import numpy as np

from ragrep.config import Config
from ragrep.ingest.chunk import chunk_all
from ragrep.ingest.embed import make_embedder
from ragrep.ingest.normalize import normalize_all
from ragrep.ingest.store import build_bm25_index, build_faiss_index, save_index
from ragrep.models import Chunk, IngestStats

log = logging.getLogger(__name__)


def _content_hash(text: str) -> str:
    """Stable content hash for embedding cache dedup."""
    return hashlib.sha256(text.encode()).hexdigest()


def _load_embed_cache(cache_path: Path, index_dir: Path) -> dict[str, np.ndarray]:
    """Load embedding cache. Bootstraps from existing FAISS index if no cache file yet."""
    if cache_path.exists():
        with open(cache_path, "rb") as f:
            cache: dict[str, np.ndarray] = pickle.load(f)
        log.info("Loaded embedding cache: %d entries", len(cache))
        return cache

    # Bootstrap from existing index (one-time migration)
    faiss_path = index_dir / "faiss.index"
    chunks_path = index_dir / "chunks.pkl"
    if not (faiss_path.exists() and chunks_path.exists()):
        return {}

    log.info("Bootstrapping embedding cache from existing index")
    idx = faiss.read_index(str(faiss_path))
    with open(chunks_path, "rb") as f:
        chunks: list[Chunk] = pickle.load(f)

    embeddings = idx.reconstruct_n(0, idx.ntotal)
    cache = {}
    for chunk, emb in zip(chunks, embeddings, strict=True):
        cache[_content_hash(chunk.content)] = emb
    log.info("Bootstrapped %d embeddings into cache", len(cache))
    return cache


def _save_embed_cache(cache_path: Path, cache: dict[str, np.ndarray]) -> None:
    """Persist embedding cache to disk."""
    cache_path.parent.mkdir(parents=True, exist_ok=True)
    with open(cache_path, "wb") as f:
        pickle.dump(cache, f)
    size_mb = cache_path.stat().st_size / 1024 / 1024
    log.info("Saved embedding cache: %d entries (%.1f MB)", len(cache), size_mb)


def ingest(config: Config, force: bool = False, source_filter: str | None = None) -> IngestStats:
    """Run the full ingestion pipeline: normalize → chunk → embed → store.

    Default: incremental — reuses cached embeddings, only embeds new/changed chunks.
    With force=True: ignores cache, re-embeds everything from scratch.
    """
    start = time.monotonic()
    raw_dir = config.raw_dir
    index_dir = config.index_dir

    # 1. Normalize
    log.info("Step 1/4: Normalizing raw data from %s", raw_dir)
    docs = normalize_all(raw_dir)

    if source_filter:
        docs = [d for d in docs if d.source == source_filter]
        log.info("Filtered to %d documents (source=%s)", len(docs), source_filter)

    if not docs:
        log.warning("No documents to index")
        return IngestStats()

    # 2. Chunk
    log.info("Step 2/4: Chunking %d documents", len(docs))
    chunks = chunk_all(docs, config.ingest.max_chunk_tokens, config.ingest.chunk_overlap_tokens)

    # 3. Embed (incremental)
    log.info("Step 3/4: Embedding %d chunks", len(chunks))

    cache_path = index_dir / "embed_cache.pkl"
    cache: dict[str, np.ndarray] = {} if force else _load_embed_cache(cache_path, index_dir)

    hashes = [_content_hash(c.content) for c in chunks]
    to_embed_idx = [i for i, h in enumerate(hashes) if h not in cache]
    n_cached = len(chunks) - len(to_embed_idx)

    if to_embed_idx:
        log.info("Embedding %d new chunks (%d cached)", len(to_embed_idx), n_cached)
        embedder = make_embedder(config.embedding.provider, config.embedding.model_name, config.embedding.device)
        new_texts = [chunks[i].content for i in to_embed_idx]
        checkpoint_path = index_dir / ".embed_checkpoint.npz"
        new_embeddings = embedder.embed_documents(
            new_texts, config.ingest.batch_size, checkpoint_path=checkpoint_path,
        )
        for j, idx in enumerate(to_embed_idx):
            cache[hashes[idx]] = new_embeddings[j]
    else:
        log.info("All %d chunks cached, no embedding needed", len(chunks))

    # Assemble full embedding matrix in chunk order
    embeddings = np.array([cache[h] for h in hashes], dtype=np.float32)
    _save_embed_cache(cache_path, cache)

    # 4. Store
    log.info("Step 4/4: Building and saving index")
    faiss_index = build_faiss_index(embeddings)
    bm25 = build_bm25_index(chunks)
    save_index(index_dir, faiss_index, chunks, bm25)

    elapsed = time.monotonic() - start

    # Compute stats
    source_counts: dict[str, int] = {}
    for doc in docs:
        source_counts[doc.source] = source_counts.get(doc.source, 0) + 1

    stats = IngestStats(
        documents=len(docs),
        chunks=len(chunks),
        sources=source_counts,
        elapsed_s=elapsed,
    )

    log.info(
        "Ingestion complete: %d docs → %d chunks (%d new, %d cached) in %.1fs. Sources: %s",
        stats.documents, stats.chunks, len(to_embed_idx), n_cached, stats.elapsed_s,
        ", ".join(f"{k}={v}" for k, v in sorted(stats.sources.items())),
    )
    return stats

File: rust/src/ingest/pipeline.rs
//! Ingest orchestrator: scrape → normalize → chunk → embed → store.

use std::collections::BTreeMap;
use std::time::Instant;

use anyhow::{Context, Result};

use crate::config::Config;
use crate::embed;
use crate::index::{bm25::Bm25, store};
use crate::ingest::{chunk, normalize};
use crate::models::IngestStats;

const CHECKPOINT_FILE: &str = ".embed_checkpoint.bin";

pub fn run(cfg: &Config, force: bool, source_filter: Option<&str>) -> Result<IngestStats> {
    let start = Instant::now();
    let raw_dir = cfg.raw_dir();
    let index_dir = cfg.index_dir();
    std::fs::create_dir_all(&index_dir)
        .with_context(|| format!("creating {}", index_dir.display()))?;

    // 1. Resolve embedder via factory
    let embedder = embed::make(&cfg.embedding.provider, &cfg.embedding.model_name)?;
    tracing::info!(
        target: "ragrep::ingest",
        "embedder ready: provider={} model={} dim={}",
        embedder.provider(),
        embedder.model(),
        embedder.dim()
    );

    // 2. Load per-(provider, model) cache (empty if --force or file absent)
    let mut cache = if force {
        embed::cache::Cache::empty(embedder.provider(), embedder.model(), embedder.dim())
    } else {
        embed::cache::load(
            &index_dir,
            embedder.provider(),
            embedder.model(),
            embedder.dim(),
        )?
    };
    tracing::info!(
        target: "ragrep::ingest",
        "embed cache: {} entries",
        cache.len()
    );

    // 3. Normalize
    tracing::info!(target: "ragrep::ingest", "step 1/4 normalize: reading {}", raw_dir.display());
    let mut docs = normalize::normalize_all(&raw_dir)?;
    if let Some(s) = source_filter {
        docs.retain(|d| d.source == s);
        tracing::info!(target: "ragrep::ingest", "filtered to {} documents (source={s})", docs.len());
    }
    if docs.is_empty() {
        tracing::warn!(target: "ragrep::ingest", "no documents to ingest");
        return Ok(IngestStats::default());
    }

    // 4. Chunk
    tracing::info!(target: "ragrep::ingest", "step 2/4 chunk: {} docs", docs.len());
    let chunks = chunk::all(
        &docs,
        cfg.ingest.max_chunk_tokens,
        cfg.ingest.chunk_overlap_tokens,
    );

    // 5. Hash + diff vs cache
    tracing::info!(target: "ragrep::ingest", "step 3/4 embed: {} chunks", chunks.len());
    let hashes: Vec<embed::cache::Hash> = chunks
        .iter()
        .map(|c| embed::cache::hash(&c.content))
        .collect();
    let to_embed_idx: Vec<usize> = hashes
        .iter()
        .enumerate()
        .filter_map(|(i, h)| if cache.contains(h) { None } else { Some(i) })
        .collect();
    let cached_count = chunks.len() - to_embed_idx.len();
    tracing::info!(
        target: "ragrep::ingest",
        "embed plan: {} new, {} cached",
        to_embed_idx.len(),
        cached_count
    );

    // 6. Embed misses
    if !to_embed_idx.is_empty() {
        let texts: Vec<&str> = to_embed_idx
            .iter()
            .map(|i| chunks[*i].content.as_str())
            .collect();
        let checkpoint = index_dir.join(CHECKPOINT_FILE);
        let new_embeddings = embedder
            .embed_documents(&texts, cfg.ingest.batch_size, Some(&checkpoint))
            .context("embedding new chunks")?;
        for (j, idx) in to_embed_idx.iter().enumerate() {
            cache.insert(hashes[*idx], new_embeddings[j].clone());
        }
    }

    // 7. Assemble full embedding matrix in chunk order
    let mut embeddings: Vec<Vec<f32>> = Vec::with_capacity(chunks.len());
    for h in &hashes {
        embeddings.push(
            cache
                .get(h)
                .cloned()
                .ok_or_else(|| anyhow::anyhow!("embedding missing for hash after embed step"))?,
        );
    }

    // 8. Save chunks + BM25 + embeddings (atomic) + cache
    tracing::info!(target: "ragrep::ingest", "step 4/4 store: writing index files");
    let bm25 = Bm25::build(chunks.iter().map(|c| c.content.as_str()));
    store::save_index(&index_dir, &chunks, &bm25, &embeddings, embedder.dim())?;
    embed::cache::save(&cache, &index_dir)?;

    let elapsed = start.elapsed().as_secs_f64();

    let mut sources: BTreeMap<String, usize> = BTreeMap::new();
    for d in &docs {
        *sources.entry(d.source.clone()).or_insert(0) += 1;
    }

    let stats = IngestStats {
        documents: docs.len(),
        chunks: chunks.len(),
        sources,
        elapsed_s: elapsed,
    };

    tracing::info!(
        target: "ragrep::ingest",
        "done: {} docs → {} chunks ({} new, {} cached) in {:.1}s",
        stats.documents,
        stats.chunks,
        to_embed_idx.len(),
        cached_count,
        stats.elapsed_s
    );
    Ok(stats)
}
